telluric/products.py

```diff
@@ -0,0 +1,838 @@
+        matched_bands = {}
+        required_bands = required_bands or cls.required_bands
+        bands = set(available_bands)
+        for dest_band in required_bands:
+            matched_bands[dest_band] = cls.match_bands(sensor_bands_info, bands, dest_band)
+        return matched_bands
+
+    @classmethod
+    def match_bands_for_legend(cls, sensor_bands_info, available_bands):
+        bands_mapping = cls.match_available_bands_to_required_bands(sensor_bands_info, available_bands)
+        bands_mapping = cls.override_bands_mapping(bands_mapping)
+        return bands_mapping
+
+    def extract_bands(self, raster, bands_mapping):
+        """return bands in required_bands if required_bands is empty it returns all bands"""
+        bands = {dest_band: self._merge_bands(bands_mapping[dest_band], raster)
+                 for dest_band in self.required_bands}
+        if self.should_convert_to_float:
+            bands = {band_name: band.astype(np.float32) for band_name, band in bands.items()}
+        return bands
+
+    def override_product_settings(self, product):
+        """this function is used in subclasses to modify products settings"""
+        return product
+
+    @classmethod
+    def override_bands_mapping(cls, bands_mapping):
+        """this function is used in subclasses to modify bands_mapping"""
+        return bands_mapping
+
+    def apply(self, sensor_bands_info, raster, bands_restriction=None, metadata=False, **kwargs):
+        """
+        Apply product calculation on the raster
+
+        :param raster: the data raster the product should apply on
+        :param bands_restriction: optional list of bands required for the processing of the product,
+            if one of this bands not exists in the raster the product should return `no data`
+        :param metadata: when true returns a named tuple with raster and additional information
+        :param kwargs: additional arguments
+        """
+        self.fits_raster_bands(raster.band_names, sensor_bands_info, bands_restriction, silent=False)
+        bands_mapping = self.match_available_bands_to_required_bands(sensor_bands_info, raster.band_names)
+        bands = self.extract_bands(raster, bands_mapping)
+        # to silence error on 0 division, which happens at nodata
+        with np.errstate(divide='ignore', invalid='ignore'):
+            array = self._apply(**bands)  # actual renderer calculation, implemented by derived class
+        used_band_names = [item for sublist in bands_mapping.values() for item in sublist]
+        array = self._force_nodata_where_it_was_in_original_raster(array, raster, used_band_names)
+
+        product_raster = raster.copy_with(image=array, band_names=self.output_bands)
+        if not metadata:
+            return product_raster
+
+        product = _Product(raster=product_raster, bands_mapping=bands_mapping, product_generator=self.to_dict())
+        product = self.override_product_settings(product)
```

Comment:
why creation of product affect `self` which is a ProductGenerator?

Reply:
it does not the product_generator subclass updated the product with its own settings